### PR TITLE
Add basic OpenReg module scaffolding with autograd

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131814
* __->__ #131708

### CODE REVIEW

**Comment by ezyang** [comment]
How does this compare to the C++ extension example that people have previously been chucking tests into?

**Comment by albanD** [comment]
As of this PR it is partially overlapping (basic scaffolding in c++).
The goal is to extend this to be a full backend though in a follow up PR which cannot be done with the current testing backend.

Once this new backend is fully up and running, it should replace the partial one we are using for testing today.

File: test/cpp_extensions/open_registration_extension/pytorch_openreg/csrc/Module.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    PyModuleDef_HEAD_INIT,
+    .m_name = "pytorch_openreg._C",
+};
+
+PyMODINIT_FUNC PyInit__C(void) {
+    PyObject* mod = PyModule_Create(&openreg_C_module);
+
+    py::object openreg_mod = py::module_::import("pytorch_openreg");
+    // Intentionally leaked
+    openreg::set_impl_registry(openreg_mod.attr("_IMPL_REGISTRY").ptr());
```
Comment (ezyang) [inline test/cpp_extensions/open_registration_extension/pytorch_openreg/csrc/Module.cpp:14]: Is it though? It sure doesn't look like you're taking an owning reference to this

File: test/cpp_extensions/open_registration_extension/test/test_openreg.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class TestOpenReg(TestCase):
+    def test_initializes(self):
+        self.assertEqual(torch._C._get_privateuse1_backend_name(), "openreg")
+
+    def test_autograd_init(self):
+        # Make sure autograd is initialized
+        torch.rand(2, requires_grad=True, device="openreg").sum().backward()
+
+        pid = os.getpid()
+        task_path = f'/proc/{pid}/task'
```
Comment (ezyang) [inline test/cpp_extensions/open_registration_extension/test/test_openreg.py:18]: Linux only

File: test/cpp_extensions/open_registration_extension/test/test_openreg.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        all_threads = psutil.Process(pid).threads()
+
+        all_thread_names = set()
+
+        for t in all_threads:
+            with open(f'{task_path}/{t.id}/comm', 'r') as file:
+                thread_name = file.read().strip()
+            all_thread_names.add(thread_name)
+
+        print(all_thread_names)
```
Comment (ezyang) [inline test/cpp_extensions/open_registration_extension/test/test_openreg.py:28]: print

**Review by ezyang — APPROVED** [review]
sure why not

File: test/cpp_extensions/open_registration_extension/pytorch_openreg/csrc/Module.cpp (hunk 1 above)
Comment (albanD) [reply test/cpp_extensions/open_registration_extension/pytorch_openreg/csrc/Module.cpp:14]: Ho yes my bad. Moving this to be borrowed for now.

**Pull request closed without merging**